File: c/bj/bj6549.c
// Solution 1: following the divide and conquer "paradigm".

#include <stdio.h>
#include <stdlib.h>

int* histogram;
int n;
long long getMaxArea(int low, int high);
long long getMaxAreaConsecutive(int low, int mid, int high);
long long getmx(long long x, long long y);
long long getmn(long long x, long long y);

int main() {
	histogram = (int*)malloc(sizeof(int) * 100001);
	long long m;
	
	while (1) {
		scanf("%d", &n);
		if (n == 0) break;

		for (int i = 0; i < n; ++i)
			scanf("%d", &histogram[i]);
		m = getMaxArea(0, n - 1);
		printf("%lld\n", m);
	}

	free(histogram);

	return 0;
}

long long getmx(long long x, long long y) {
	return (x > y) ? x : y;
}

long long getmn(long long x, long long y) {
	return (x < y) ? x : y;
}

long long getMaxArea(int low, int high) {
	int mid;
	mid = (low + high) / 2;
	if (low == high) return histogram[low];

	long long left = getMaxArea(low, mid);
	long long right = getMaxArea(mid + 1, high);
	long long whole = getMaxAreaConsecutive(low, mid, high);

	long long mx = getmx(whole, getmx(left, right));

	return mx;
}

long long getMaxAreaConsecutive(int low, int mid, int high) {
	long long len = 0;
	long long leftward = mid;
	long long leftval = histogram[leftward];
	long long rightward = mid + 1;
	long long rightval = histogram[rightward];
	long long result = 2 * getmn(leftval, rightval);

	while (low < leftward && rightward < high) {
		if (leftval <= rightval) {
			if (rightval <= histogram[++rightward]) {
				result = getmx(rightval * (rightward - leftward), getmx(result, leftval * (rightward - leftward + 1)));
			} else {
				rightval = histogram[rightward];
				result = getmx(result, getmn(rightval, leftval) * (rightward - leftward + 1));
			}
		} else {
			if (leftval <= histogram[--leftward]) {
				result = getmx(leftval * (rightward - leftward), getmx(result, rightval * (rightward - leftward + 1)));
			} else {
				leftval = histogram[leftward];
				result = getmx(result, getmn(rightval, leftval) * (rightward - leftward + 1));
			}
		}
	}
	
	while (low < leftward) {
		leftval = getmn(leftval, histogram[--leftward]);
		if (leftval <= rightval)
			result = getmx(result, leftval * (rightward - leftward + 1));
		else
			result = getmx(result, getmx(rightval * (rightward - leftward + 1), leftval * (rightward - leftward)));
	}
	while (rightward < high) {
		rightval = getmn(rightval, histogram[++rightward]);
		if (rightval <= leftval)
			result = getmx(result, rightval * (rightward - leftward + 1));
		else
			result = getmx(result, getmx(leftval * (rightward - leftward + 1), rightval * (rightward - leftward)));
	}

	return result;
}

File: java/src/baekjoon/BJ6549.java
package baekjoon;

import java.io.*;
import java.util.StringTokenizer;

// This version follows the divide-and-conquer paradigm; it is less efficient than
// the solution that uses the stack data structure.
public class BJ6549 {

    private static long[] histogram = new long[100001];

    private static long getAreaConsecutive(int low, int high, int mid) {
        int leftward = mid;
        int rightward = mid;
        long height = histogram[mid];
        long max = height;

        while (low < leftward && rightward < high) {
            if (histogram[leftward - 1] < histogram[rightward + 1]) {
                rightward++;
                height = Math.min(histogram[rightward], height);
            } else {
                leftward--;
                height = Math.min(histogram[leftward], height);
            }
            max = Math.max(max, height * (rightward - leftward + 1));
        }

        while (rightward < high) {
            rightward++;
            height = Math.min(height, histogram[rightward]);
            max = Math.max(max, height * (rightward - leftward + 1));
        }

        while (leftward > low) {
            leftward--;
            height = Math.min(height, histogram[leftward]);
            max = Math.max(max, height * (rightward - leftward + 1));
        }

        return max;
    }

    private static long getArea(int low, int high) {
        // Base case: segment only contains one bar.
        if (low == high) return histogram[low];

        int mid = (low + high) / 2;

        // max area of the left half
        long left = getArea(low, mid);
        // max area of the right half
        long right = getArea(mid + 1, high);
        // max area of the whole
        long whole = getAreaConsecutive(low, high, mid);

        long max = Math.max(whole, Math.max(left, right));

        return max;
    }

    private static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        while (n != 0) {
            for (int i = 0; i < n; ++i) {
                histogram[i] = Integer.parseInt(st.nextToken());
            }

            sb.append(getArea(0, n - 1)).append('\n');

            st = new StringTokenizer(br.readLine());
            n = Integer.parseInt(st.nextToken());
        }

        bw.write(sb.toString());
        bw.flush();
    }

    public static void main(String[] args) throws IOException {
        solve();
    }

}
